File: Data Structures & Algorithms/implement-prefix-tree/submission-3.java


class Node{
    HashMap<String,Node> children;
    boolean is_word_end;
    public Node(){
        this.children = new HashMap<>();
        this.is_word_end = false;
    }
}

class PrefixTree {

    Node root;
    public PrefixTree() {
         this.root = new Node();
    }

    public void insert(String word) {
        Node copy = this.root;
        for(int i=0;i<word.length();i++){
            char curChar = word.charAt(i);
            String curString = new String(new char[]{curChar});
            if(!copy.children.containsKey(curString)){
                copy.children.put(curString, new Node());
            }
            copy = copy.children.get(curString);
        }
        copy.is_word_end = true;
    }

    public boolean search(String word) {
        Node node = this.root;

        for(int i=0;i<word.length();i++){
            char curChar = word.charAt(i);
            String curString = new String(new char[]{curChar});
            // System.out.println(" i "+i+" "+node.children);
            if(!node.children.containsKey(curString)){
                return false;
            }
            node = node.children.get(curString);
        }

        return true && node.is_word_end;

    }

    public boolean startsWith(String word) {

        Node node = this.root;

        for(int i=0;i<word.length();i++){
            char curChar = word.charAt(i);
            String curString = new String(new char[]{curChar});
            if(!node.children.containsKey(curString)){
                return false;
            }
            node = node.children.get(curString);
        }

        return true;
        
    }
}

File: Data Structures & Algorithms/extra-characters-in-a-string/submission-3.py
from collections import defaultdict

class Node:
    def __init__(self) -> None:
        self.children = {}
        self.is_word_end = False

class Trie:

    def __init__(self) -> None:
        self.root = Node()
    
    def addWord(self,word):
        Node1 = self.root
        for i in word:
            if(i not in Node1.children):
                Node1.children[i] = Node()
            Node1 = Node1.children[i]
        Node1.is_word_end = True

    def Search(self,word):
        Node = self.root
        for i in word:
            if(i in Node.children):
                Node = Node.children[i]
            else:
                return False
        return True and Node.is_word_end


class Solution:
    def minExtraChar(self, s: str, dictionary: List[str]) -> int:

        trie = Trie()
        for i in dictionary:
            trie.addWord(i)

        DP = {}

        def DFS(index):
            if(index == len(s)):
                return 0
            if(index in DP):
                return DP[index]

            res = 1+DFS(index+1)
            for i in range(index,len(s)):
                if(trie.Search(s[index:i+1])):
                    res = min(res,DFS(i+1))
            
            DP[index] = res
            return res
        
        return DFS(0)

        